internal/modules/modules.go

```diff
@@ -10,123 +10,174 @@ package modules
-	for _, d := range deps {
-		if shouldSkipModule(modules, d) {
-			continue
+			// check if the current module already is this version
+			// IDEA(jaredallard): In the future we probably want to see if
+			// the past version _also_ satisfies the new constraints, and if so
+			// skip re-resolving the module.
+			if version == resolved[importPath].Version {
+				modulesToResolve = modulesToResolve[1:]
+				continue
+			}
 		}
 
-		// create a module struct for this module, this resolves the latest version if
-		// the version wasn't set.
-		m, err := New(ctx, sm.Replacements[d.Name], d)
+		m, err := New(ctx, uri, &configuration.TemplateRepository{
+			Name:    importPath,
+			Channel: rm.conf.Channel,
+			Version: version,
+		})
 		if err != nil {
-			return errors.Wrapf(err, "failed to create dependency %q", d.Name)
+			return nil, err
 		}
 
-		// prefetch the manifest and FS
 		mf, err := m.Manifest(ctx)
 		if err != nil {
-			return errors.Wrapf(err, "failed to parse manifest %q", d.Name)
+			return nil, err
 		}
-		modules[d.Name] = &consideredModule{m, d.Prerelease}
 
-		// if we have dependencies, download those now
-		if len(mf.Modules) != 0 {
-			if err := getModulesForService(ctx, sm, mf.Modules, modules); err != nil {
-				return errors.Wrapf(err, "failed to process dependency of %q", d.Name)
-			}
+		// add the dependencies of this module to the stack to be resolved
+		for i := range mf.Modules {
+			modulesToResolve = append(modulesToResolve, resolveModule{
+				conf:   mf.Modules[i],
+				parent: importPath,
+			})
+		}
+
+		// set the module on our resolved module
+		resolved[importPath].Module = m
+
+		// resolve the next module
+		modulesToResolve = modulesToResolve[1:]
+	}
+
+	// convert the resolved modules to a list of modules
+	modules := make([]*Module, 0, len(resolved))
+	for _, m := range resolved {
+		modules = append(modules, m.Module)
+	}
+	return modules, nil
+}
+
+// getLatestModuleForConstraints returns the latest module that satisfies the provided constraints
+func getLatestModuleForConstraints(ctx context.Context, uri string, token cfg.SecretData,
+	m *resolveModule, resolved map[string]*resolvedModule) (string, error) {
+	constraints := resolved[m.conf.Name].constraints
+	if len(constraints) == 0 {
+		constraints = make([]constraint, 0)
+	}
+
+	// if we have a constraint, use it to resolve the version
+	if m.conf.Version != "" {
+		constraints = append(constraints, constraint{
+			str:          m.conf.Version,
+			parentModule: m.parent,
+		})
+		resolved[m.conf.Name].constraints = constraints
+	}
+
+	constraintsStr := make([]string, len(constraints))
+	for i := range constraints {
+		constraintsStr[i] = constraints[i].str
+	}
+
+	v, err := resolver.Resolve(ctx, token, &resolver.Criteria{
+		URL:         uri,
+		Channel:     m.conf.Channel,
+		Constraints: constraintsStr,
+	})
+	if err != nil {
+		errorString := ""
+		for i := range constraints {
```

Comment:
I'll want to expose the resolution information in a different format than the error string, but that'll increase the scope of this work a decent amount, so for now this is why this is going in the error string.

If future me gets mad about this, I'm sorry 😢